lib/stack/src/map_err.rs

```diff
@@ -0,0 +1,72 @@
+pub fn layer<E, M>(map_err: M) -> Layer<M>
+where
+    M: MapErr<E>,
+{
+    Layer(map_err)
+}
+
+pub(super) fn stack<T, S, M>(inner: S, map_err: M) -> Stack<S, M>
+where
+    S: super::Stack<T>,
+    M: MapErr<S::Error>,
+{
+    Stack {
+        inner,
+        map_err,
+    }
+}
+
+pub trait MapErr<Input> {
```

Comment:
I'm assuming that adding this trait, rather than just using `Fn(Input) -> Output` directly, is to allow structs to implement a stateful `MapErr`? It might be good to have a comment explaining that...

Reply:
It's really so that it's possible to have a named type. I think it's an anti-pattern to expose `Fn` generics in public interfaces.